### PR TITLE
Bugfix/wificlient msys 188

### PR DESCRIPTION
**Pull request by nhjschulz**
Make WifiClient.h platform neutral
Compile with platform specific implementations for POSIX and Windows (winsock2).

### CODE REVIEW

File: platformio.ini [hunk 1]
```diff
@@ -118,6 +119,7 @@ lib_ignore =
     MainTestNative
 extra_scripts =
     ./scripts/webots_launcher.py
+    pre:./scripts/add_os_specific_build_flags.py
```
Comment (BlueAndi) [inline platformio.ini:122]: Shouldn't it be added to the target:Test section too?

File: lib/HALSim/src/Board.cpp [hunk 2]
```diff
@@ -68,6 +74,17 @@ bool Board::init()
...
     bool isReady = true;
 
+#if _WIN32
+    WORD wVersionRequested = MAKEWORD(2,2);
+    WSADATA wsaData;
+    int result = WSAStartup(wVersionRequested, &wsaData);
+
+    if (0 != result)
+    {
+        LOG_ERROR("WSAStartup error %d", result);
```
Comment (BlueAndi) [inline lib/HALSim/src/Board.cpp:84]: Shouldn't the return value for init() be influenced?

**Review by BlueAndi — CHANGES_REQUESTED** [review]
HALTest is not considered, but should.

File: lib/WiFiNative/src/WifiClientPosix.inl [hunk 3]
```diff
@@ -155,17 +155,16 @@ void WiFiClient::stop()
...
+        auto iterSock = gConnections.find(this);
 
-        if (gSockets.end() != iterSock)
+        if (gConnections.end() != iterSock)
         {
             if (0 != ::close(iterSock->second.fd))
             {
                 LOG_ERROR("%s:%s", "close", strerror(errno));
             }
+            gConnections.erase(iterSock);
```
[the remark this answers is not in the document]
Comment (nhjschulz) [reply lib/WiFiNative/src/WifiClientPosix.inl:166]: Added (void) upfront.

File: lib/WiFiNative/src/WifiClientPosix.inl [hunk 4]
```diff
@@ -111,15 +111,15 @@ uint8_t WiFiClient::connect(const IPAddress& addr, uint16_t port)
         serverAddr.sin_addr.s_addr = htonl(addr);
         serverAddr.sin_port        = htons(port);
 
-        if (0 == ::connect(socket, reinterpret_cast<struct sockaddr*>(&serverAddr), sizeof(serverAddr)))
+        if (0 == ::connect(socketFd, reinterpret_cast<struct sockaddr*>(&serverAddr), sizeof(serverAddr)))
         {
             const int one = 1;
 
-            if (-1 == ::setsockopt(socket, SOL_TCP, TCP_NODELAY, &one, sizeof(one)))
+            if (-1 == ::setsockopt(socketFd, SOL_TCP, TCP_NODELAY, &one, sizeof(one)))
```
[the remark this answers is not in the document]
Comment (nhjschulz) [reply lib/WiFiNative/src/WifiClientPosix.inl:118]: ok

File: lib/WiFiNative/src/WifiClientWinsock.inl [hunk 5]
```diff
@@ -105,20 +105,20 @@ uint8_t WiFiClient::connect(const IPAddress& addr, uint16_t port)
         serverAddr.sin_addr.s_addr = htonl(addr);
         serverAddr.sin_port        = htons(port);
 
-        if (0 == ::connect(socket, reinterpret_cast<struct sockaddr*>(&serverAddr), sizeof(serverAddr)))
+        if (0 == ::connect(socketHandle, reinterpret_cast<struct sockaddr*>(&serverAddr), sizeof(serverAddr)))
         {
             const int one     = 1;
```
[the remark this answers is not in the document]
Comment (nhjschulz) [reply lib/WiFiNative/src/WifiClientWinsock.inl:110]: ok

File: lib/WiFiNative/src/WifiClientPosix.inl [hunk 6]
```diff
@@ -111,15 +111,15 @@ uint8_t WiFiClient::connect(const IPAddress& addr, uint16_t port)
         serverAddr.sin_addr.s_addr = htonl(addr);
         serverAddr.sin_port        = htons(port);
 
-        if (0 == ::connect(socket, reinterpret_cast<struct sockaddr*>(&serverAddr), sizeof(serverAddr)))
+        if (0 == ::connect(socketFd, reinterpret_cast<struct sockaddr*>(&serverAddr), sizeof(serverAddr)))
         {
             const int one = 1;
```
[the remark this answers is not in the document]
Comment (nhjschulz) [reply lib/WiFiNative/src/WifiClientPosix.inl:116]: now             const int enableOptVal = 1;

File: platformio.ini [hunk 7]
```diff
@@ -99,6 +99,7 @@ build_flags =
     ${mode:selected_native.build_flags}
     -std=c++11
     -D TRANSPORT_USE_TCP=1
+    -D_USE_MATH_DEFINES
```
[the remark this answers is not in the document]
Comment (nhjschulz) [reply platformio.ini:102]: removed line, was dublicate define

File: lib/HALSim/src/Board.cpp (hunk 2 above)
Comment (nhjschulz) [reply lib/HALSim/src/Board.cpp:84]: will now caus init() to fail.

File: lib/WiFiNative/src/WifiClientPosix.inl [hunk 8]
```diff
@@ -0,0 +1,241 @@
...
+    {
+        auto iterSock = gConnections.find(this);
+
+        if (gConnections.end() != iterSock)
+        {
+            if (0 != ::close(iterSock->second.fd))
+            {
+                LOG_ERROR("%s:%s", "close", strerror(errno));
+            }
+            (void)Connections.erase(iterSock);
```
Comment (BlueAndi) [inline lib/WiFiNative/src/WifiClientPosix.inl:166]: Same iterator is used with different containers 'gConnections' and 'Connections'.

File: lib/HALSim/src/Board.cpp [hunk 9]
```diff
@@ -32,10 +32,16 @@
 /******************************************************************************
  * Includes
  *****************************************************************************/
+#if _WIN32
```
Comment (BlueAndi) [inline lib/HALSim/src/Board.cpp:35]: Shouldn't it be "#ifdef ...." at all places?

**Review by BlueAndi — CHANGES_REQUESTED** [review]

File: lib/WiFiNative/src/WifiClientPosix.inl (hunk 8 above)
Comment (nhjschulz) [reply lib/WiFiNative/src/WifiClientPosix.inl]: typo :(

File: lib/HALSim/src/Board.cpp (hunk 9 above)
Comment (nhjschulz) [reply lib/HALSim/src/Board.cpp]: Yes, it behaves the same but #if <maybe undef> isn't MISRA compliant - changed.

**Review by BlueAndi — CHANGES_REQUESTED** [review]

**Pull request merged**